mobile/src/main/java/org/openhab/habdroid/util/HttpClient.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2010-2016, openHAB.org and others.
+ *
+ *   All rights reserved. This program and the accompanying materials
+ *   are made available under the terms of the Eclipse Public License v1.0
+ *   which accompanies this distribution, and is available at
+ *   http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.openhab.habdroid.util;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+import android.security.KeyChain;
+import android.security.KeyChainException;
+import android.support.annotation.VisibleForTesting;
+import android.util.Log;
+
+import java.net.Socket;
+import java.security.KeyStore;
+import java.security.Principal;
+import java.security.PrivateKey;
+import java.security.SecureRandom;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.TrustManagerFactory;
+import javax.net.ssl.X509KeyManager;
+import javax.net.ssl.X509TrustManager;
+
+import de.duenndns.ssl.MemorizingTrustManager;
+import okhttp3.Call;
+import okhttp3.Credentials;
+import okhttp3.HttpUrl;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+
+public abstract class HttpClient {
+    private static final String TAG = HttpClient.class.getSimpleName();
+
+    private final HttpUrl mBaseUrl;
+    private final Map<String, String> headers = new HashMap<>();
+
+    private OkHttpClient mClient;
+    private HostnameVerifier mDefaultHostnameVerifier;
+    private SSLSocketFactory mDefaultSocketFactory;
+
+    private static final List<String> SSL_RELEVANT_KEYS = Arrays.asList(
+            Constants.PREFERENCE_SSLHOST, Constants.PREFERENCE_SSLCERT,
+            Constants.PREFERENCE_SSLCLIENTCERT
+    );
+
+    protected HttpClient(Context context, String baseUrl) {
+        this(context, PreferenceManager.getDefaultSharedPreferences(context), baseUrl);
+    }
+
+    protected HttpClient(Context context, SharedPreferences prefs, String baseUrl) {
+        final Context appContext = context.getApplicationContext();
+        mClient = new OkHttpClient.Builder().build();
+        mBaseUrl = baseUrl != null ? HttpUrl.parse(baseUrl) : null;
+
+        mDefaultHostnameVerifier = mClient.hostnameVerifier();
+        mDefaultSocketFactory = mClient.sslSocketFactory();
+
+        applySslProperties(appContext, prefs);
+        prefs.registerOnSharedPreferenceChangeListener((prefsInstance, key) -> {
```

Comment:
Almost missed this one; that's the one item left to fix: With registering the listener we'd need to unregister it at some point. We could either add a `dispose()` (or similar) method to HttpClient and Connection which unregisters the listener, or we could call `applySslProperties()` from outside (e.g. ConnectionFactory) on pref changes.

@mueller-ma @FlorianSW Which variant would you consider more elegant? Both have the implication of the user of the class doing The Right Thing[TM] :-/

Reply:
Thinking about it it's probably not a big issue, since we're going to kill off the SSL options soon anyway. I already have prepared a patch for that; will submit once this PR + MTM update is done.

Reply:
The instance of HttpClient should be idempotent when it comes to it's settings (one reason, why the object shouldn't be cached in client/caller code). When the settings for preferences, used by the http client, changes, a new instance should be created and provided to the calling code. That all should be handled by the connection object (which already needs to track some settings, iirc).

A second thought: I'm not sure, if I like it, that the HttpClient takes a SharedPrefs object here. It would be more explicit if the actual values would be passed to the instance, as it would make clear for other code, what it really expects, instead of taking something, which is somewhere in the settings.